src/checkers/inference/solver/frontend/ConstraintSerializer.java

```diff
@@ -0,0 +1,113 @@
+package checkers.inference.solver.frontend;
+
+import java.lang.reflect.Constructor;
+
+import checkers.inference.model.CombVariableSlot;
+import checkers.inference.model.CombineConstraint;
+import checkers.inference.model.ComparableConstraint;
+import checkers.inference.model.ConstantSlot;
+import checkers.inference.model.EqualityConstraint;
+import checkers.inference.model.ExistentialConstraint;
+import checkers.inference.model.ExistentialVariableSlot;
+import checkers.inference.model.InequalityConstraint;
+import checkers.inference.model.PreferenceConstraint;
+import checkers.inference.model.RefinementVariableSlot;
+import checkers.inference.model.Serializer;
+import checkers.inference.model.SubtypeConstraint;
+import checkers.inference.model.VariableSlot;
+import checkers.inference.solver.backend.BackEndType;
+import checkers.inference.solver.backend.BackEndType.BackEndTypeEnum;
+
+/**
+ * ConstraintSerializer is a deliverer that can delegate each type of
+ * constraint/slot to the real serializer that is specified by user.
+ * 
+ * The costume serialization can be achieved by overriding the serialize methods
+ * in this class.
+ * 
+ * @author jianchu
+ *
+ * @param <S> Encoding type for slot.
+ * @param <T> Encoding type for constraint.
+ */
+public class ConstraintSerializer<S, T> implements Serializer<S, T> {
+
+    private Serializer<S, T> realSerializer;
+
+    @SuppressWarnings("unchecked")
+    public ConstraintSerializer(BackEndType backEndType, Lattice lattice) {
+        try {
+            String backEndPath = backEndType.getFullyQualifiedName();
+            if (backEndType.getBackEndType().equals(BackEndTypeEnum.Lingeling)) {
```

Comment:
I would suggest to de-coupling the general `ConstraintSerializer` with a specific BackEndType. But how to achieve this need more discussion.

Also, I have a more general question on the design level: what is the initial motivation to highly decoupling BackEnd and Serializer?

It seems each BackEnd always use a specific kind of Serializer, and this is an implementation detail hidden from user (i.e. user doesn't have an argument to indicate which serializer should be used). So to me it seems Serializer should be a component of BackEnd.

And the motivation of my question is, I meet a problem when developing BackEnds in JaCop and Z3:

In JaCop and Z3, there is a common object that responsible for both creating variables, constraints, and adding constraints to solver. In other words, BackEnd and Serializer have to share this common object. However, in our current design, it doesn't allow me to do this, as Serializer and BackEnd are highly decoupled (they initialized separately by Solver, and only communicate with each other through interface).

So my question is, can we make a stronger coupling relationship between Serializer and BackEnd? Namely, make Serializer as an inner component of BackEnd, and each BackEnd define their own Serializer. By this way I can have more flexible interaction between BackEnd and Serializer, and reduce the burden of Solver to maintaining both BackEnd and Serializer. Also, if we choose this way, there is no need of a `ConstraintSerializer`.